geonode/resource/manager.py

```diff
@@ -558,7 +561,10 @@ def remove_permissions(self, uuid: str, /, instance: ResourceBase = None) -> boo
                 _resource.set_processing_state(enumerations.STATE_PROCESSED)
         return False
 
-    def set_permissions(self, uuid: str, /, instance: ResourceBase = None, owner: settings.AUTH_USER_MODEL = None, permissions: dict = {}, created: bool = False) -> bool:
+    def set_permissions(
+        self, uuid: str, /, instance: ResourceBase = None, owner: settings.AUTH_USER_MODEL = None,
+        permissions: dict = {}, created: bool = False
+    ) -> bool:
```

Comment:
The argument `geofence_uow: GeofenceLayerRulesUnitOfWork = None` was removed of this method.

Inside the `atomic` block of this method, Geofence is modified by:
* `self.remove_permissions(uuid, instance=_resource)`
* `self._concrete_resource_manager.set_permissions(uuid, instance=_resource, owner=owner, permissions=permissions, created=created)`

Inside each one of then there is still a atomic GeofenceLayerRulesUnitOfWork context block, but without an "outer" geonfence_uow being passed down, the following scenario is still possible:
* Geofence has X rules.
* `remove_permissions` is successfully executed. Now Geofence has 0 rules.
* `self._concrete_resource_manager.set_permissions` raises an exception, so some new rules aren't created. The rollback inside the concrete obj will restore the previous state of 0 rules.
* In this case, the Geofence data won't be restored to X rules as it was before.

Reply:
@italogsfernandes that's true but in this architecture the generic manager cannot be aware of the specific GeoFence/GeoServer stuff.
We should probably generalize here the concept of security rule and Rule Unit of Work and use the generic class to pass the rules to the concrete manager.
It is a bit out of context for the moment, I prefer to remove the rules if the object is in an inconsistent state and throw an exception instead.

Reply:
Ok